awslabs/aws-lambda-powertools-typescript
# feat(build): publish lib as a Lambda Layer

## Description of your changes

This PR aims to add the capability to publish safely AWS Lambda PowerTools for TypeScript.

TODO list:
- [x] Create the publisher code 
- [ ] Make https://github.com/aws-samples/cdk-lambda-powertools-python-layer/pull/52 merged
- [x] Add e2e for all runtimes
- [x] decide what we do for Node12 : #899
- [x] Add publication flow to release github workflow
- [x] Figure out the public access of the layer
- [x] Update the doc
- [ ] Add https://pypi.org/project/mkdocs-simple-hooks/ to update automatically the doc ... (python should do this soon so we can wait for their code :) )
- [ ] Add SAM and CDK examples

### How to verify this change

#### Test layer publisher

Build layer manually in your account:
```sh
cd layer-publisher
npm ci
npm run cdk deploy
```

Then you can use the layer shown as cfn output.

Or you can use the e2e tests that will:
1. Create a public layer in your account called `e2e-tests-layer-nodejs`
2. Create a Lambda function consuming the previously created layer
3. Check the outputs logs of the function checking no error is thrown

#### Test in github action

check https://github.com/flochaz/aws-lambda-powertools-typescript/actions/runs/2718610199

### Diff with Powertools for Python version

The only difference with Python version for publishing layer is the way we do e2e tests:
* Python is launching a `Canary` stack launching a Lambda Function (as a CustomResource) which consume the published layer in the same CDK App. This canary function deployed in their beta and prod account will use the layer and throw exception if any issue found making the stack deployment fail.
* TypeScript version is doing E2E testing the same way we were doing for npm: `layer-publisher` cdk app contains a `Jest` test that will  
  1. Create a public layer in the test account called `e2e-tests-layer-nodejs`
  2. Create a Lambda function consuming the previously created layer
  3. Check the outputs logs of the function checking no error is thrown

### Related issues, RFCs

**Issue number:** #826

### PR status

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

## Checklist


### Breaking change checklist


---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
